### PR TITLE
Cargo clippy refactoring

### PR DESCRIPTION
**Pull request by oli-obk**
Moving away from clippy as a plugin to clippy as a driver

Also: no more hacky `cargo-clippy` calling itself!

This is a prerequisite to getting clippy's test suite into Rust's CI

### CODE REVIEW

**Comment by Manishearth** [comment]
WooooooOOOOOOOOoooooooOOOOOOOooooooOOOOOoooooOOOOOoooooOOOOOoooooOOOOOooooOOOOooooOOOOooo 🎊

File: Cargo.toml [hunk 1]
```diff
@@ -29,6 +29,11 @@ name = "cargo-clippy"
 test = false
 path = "src/main.rs"
 
+[[bin]]
+name = "clippy-driver"
```
Comment (Manishearth) [inline Cargo.toml]: yesssss this is much better kinda feel like we should have done this on the get go

File: src/driver.rs [hunk 2]
```diff
@@ -0,0 +1,190 @@
...
+impl ClippyCompilerCalls {
+    fn new(run_lints: bool) -> Self {
+        Self {
+            default: RustcDefaultCalls,
+            run_lints: run_lints,
+        }
+    }
+}
+
+impl<'a> CompilerCalls<'a> for ClippyCompilerCalls {
```
Comment (Manishearth) [inline src/driver.rs]: Was any code changed here? It looks ok to me, but it's unclear what was added/removed.

File: tests/run-pass/conf_whitelisted.rs [hunk 3]
```diff
@@ -1,4 +0,0 @@
-#![feature(plugin)]
-#![plugin(clippy(conf_file="./tests/auxiliary/conf_whitelisted.toml"))]
```
Comment (mcarton) [inline tests/run-pass/conf_whitelisted.rs]: Is overriding the conf file path not supported anymore?

**Comment by llogiq** [comment]
Great work so far! :+1:

File: src/driver.rs (hunk 2 above)
Comment (oli-obk) [reply src/driver.rs]: No. I should have probably duplicated the file in a separate commit before removing things...

File: tests/run-pass/conf_whitelisted.rs (hunk 3 above)
Comment (oli-obk) [reply tests/run-pass/conf_whitelisted.rs]: Still supported. Just not testable anymore... actually. .. run pass could be a regular rust integration test. I'll check

**Comment by llogiq** [comment]
I think you also need to copy `clippy-driver`before calling `cargo-clippy` in `.travis.yml`, line 30.

**Comment by oli-obk** [comment]
I'm on mobile, and github is unusable on mobile. Can you edit it please?

**Comment by llogiq** [comment]
I'm also in mobile right now. Will try to get to my PC soon.

**Comment by llogiq** [comment]
Ok, let's see if this works.

**Comment by llogiq** [comment]
Now we need a rustup...

**Comment by Manishearth** [comment]
already on it, just needs to wait on travis

-Manish Goregaokar

**Comment by oli-obk** [comment]
🎉

**Comment by llogiq** [comment]
🎉

**Pull request merged**